bundles/edu.kit.kastel.sdq.eclipse.grading.view/src/edu/kit/kastel/eclipse/grading/view/commands/AddAnnotationDialog.java

```diff
@@ -0,0 +1,158 @@
+import org.eclipse.jface.viewers.LabelProvider;
+import org.eclipse.jface.viewers.TableViewer;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.events.KeyAdapter;
+import org.eclipse.swt.events.KeyEvent;
+import org.eclipse.swt.events.SelectionAdapter;
+import org.eclipse.swt.events.SelectionEvent;
+import org.eclipse.swt.layout.GridData;
+import org.eclipse.swt.layout.GridLayout;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Control;
+import org.eclipse.swt.widgets.Shell;
+import org.eclipse.swt.widgets.Text;
+
+import edu.kit.kastel.eclipse.common.api.model.IMistakeType;
+import edu.kit.kastel.eclipse.grading.view.controllers.AssessmentViewController;
+
+// One could also use the FilteredResourcesSelectionDialog (used e.g. for the buit-in Open Type command)
+// but I don't like the UX of this dialog, it takes a comparatively long time to open
+// and adding the shift-click listener is hard (I think)
+public class AddAnnotationDialog extends Dialog {
+	private final AssessmentViewController controller;
+	private TableViewer displayList;
+	private AnnotationFilter filter;
+	
+	private IMistakeType selectedMistake;
+	private boolean customMessageWanted;
+
+	public AddAnnotationDialog(Shell parentShell, AssessmentViewController controller) {
+		super(parentShell);
+		this.controller = controller;
+	}
+	
+	public Optional<IMistakeType> getSelectedMistake() {
+		return Optional.ofNullable(this.selectedMistake);
+	}
+	
+	public boolean isCustomMessageWanted() {
+		return this.customMessageWanted;
+	}
+	
+	@Override
+	protected Control createDialogArea(Composite parent) {
+		Composite container = (Composite) super.createDialogArea(parent);
+		container.setSize(SWT.DEFAULT, 200);
+		
+		GridLayout layout = new GridLayout();
+		container.setLayout(layout);
+		
+		this.createSearchField(container);
+		this.createAnnotationList(container);
+		
+		return container;
+	}
+	
+	@Override
+	protected void configureShell(Shell shell) {
+		super.configureShell(shell);
+		shell.setText("Add Annotation");
+	}
+	
+	@Override
+	protected void createButtonsForButtonBar(Composite parent) {
+		// Removes the ok and close buttons
+	}
+	
+	private void updateDisplayList(String filter) {
+		this.filter.setFilterString(filter);
+		this.displayList.refresh();
+		this.displayList.getTable().setSelection(0);
+	}
+	
+	private void createSearchField(Composite container) {
+		Text inputField = new Text(container, SWT.SINGLE | SWT.BORDER);
+		inputField.addTraverseListener(e -> {
+			if (e.detail == SWT.TRAVERSE_RETURN) {
+				processMistakeSelection(isShiftPressed(e.stateMask));
+				this.okPressed();
+			} else if (e.detail == SWT.TRAVERSE_ESCAPE) {
+				this.cancelPressed();
+			}
+		});
+		inputField.addKeyListener(new KeyAdapter() {
+			@Override
+			public void keyPressed(KeyEvent e) {
+				int index = displayList.getTable().getSelectionIndex();
+				if (e.keyCode == SWT.ARROW_DOWN) {
+					displayList.getTable().select(index + 1);
```

Comment:
what happens if index out of bounds?

Reply:
select ingnores the index; see [the documentation](https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fswt%2Fwidgets%2FTable.html&anchor=select(int))